batches: use "keyword" as default pattern type

stefanhengl opened the pull request:
This is part of the Keyword GA Project.

Batch Changes uses Sourcegraph queries to define the list of repositories on which the batch change will run.

With this change we default to pattern type "keyword" instead of "standard". 

To make this a backward compatible change, we also introduce a version identifier to batch specs. Authors can specify `version: 2` in the spec, in which case we default to pattern type "keyword". Existing specs (without a specified version) and specs with `version: 1` will keep using pattern type "standard".

<img width="1177" alt="Screenshot 2024-07-08 at 14 51 27" src="https://github.com/sourcegraph/sourcegraph/assets/26413131/9f1911a1-bbf1-4ddf-a9f0-fe54d1b85bdd">

Notes:
- Corresponding doc update [PR](https://github.com/sourcegraph/docs/pull/477)
- We don't have a query input field, but instead the query is defined in a batch spec YAML. It didn't feel right to edit the YAML and append "patternType: " on save, which is what we do for Code Monitors and Insights.
- I misuse the pattern type query parameter to effectively override the version. Once we introduce "V4" we should come back here and clean up. I left a TODO in the code.

Test plan:
- New and updated unit tests
- manual testing
  - new batch changes use `version: 2` by default.
  - using an unsupported version returns an error
  - I ran various "on:" queries to verify that version 2 uses keyword search and version 1 uses standard search.

stefanhengl commented:
@eseliger I am not very familiar with the Batch Changes code, so I am looking for some guidance. Do you think this approach makes sense? Especially these 2 points:
> - We don't have a query input field, but instead the query is defined in a batch spec YAML. It didn't feel right to edit the YAML and append "patternType: " on save, which is what we do for Code Monitors and Insights.
> - I hardcoded "keyword" and didn't let the pattern type depend on the settings. Several users might interact with the spec and it would be confusing if the list of repos depended on the user.

Are there any other places in Batch Changes where users can define Sourcegraph queries?

jtibshirani reviewed (commented):
This looks good to me (although I am far from a batch changes expert!) It makes sense to use the same default for every user, as batch changes are shared, similar to what we do for notebooks.

You could argue that existing batch changes should use the previous default patterntype `standard` 🤔 That way, when a user reruns the same batch change after the 5.5 upgrade, it will pull up the same repositories. Curious how important you think it would be to implement that (which would require storing patterntypes in the DB?) versus just keeping it simple with an approach like this. Clarifying: I am not saying "we definitely need to do this", genuinely asking a question!

eseliger commented:
So this means patternType is always overwritten to "keyword"? Or can you still specify the patternType inline?

I think Julies concern is valid - this would change the behavior of existing batch specs 🤔 I'm not sure how important that property is, but it should probably at least be well called out somewhere. 

Can we also make sure the examples in `client/web/src/enterprise/batches/batch-spec/edit/library/*.yaml` still work? 

---

> Are there any other places in Batch Changes where users can define Sourcegraph queries?

I think that'd be it.

stefanhengl commented:
> So this means patternType is always overwritten to "keyword"? Or can you still specify the patternType inline?

You can still specify it inline. We just take keyword if the user didn't specify anything.
 
> Can we also make sure the examples in `client/web/src/enterprise/batches/batch-spec/edit/library/*.yaml` still work?

Oops, I don't know about those. I will take a look.

stefanhengl commented on client/web/src/enterprise/batches/batch-spec/edit/library/hello-world.batch.yaml:
```diff
@@ -1,3 +1,4 @@
+version: v2
```
This example is taken as the default when opening a new batch request. By updating the example we nudge users towards using v2.

stefanhengl commented on internal/batches/service/workspace_resolver.go:
```diff
@@ -483,12 +496,12 @@ const findDirectoriesInReposConcurrency = 10
+func (wr *workspaceResolver) FindDirectoriesInRepos(ctx context.Context, fileName string, batchSpecVersion string, repos ...*RepoRevision) (map[repoRevKey][]string, error) {
 	findForRepoRev := func(repoRev *RepoRevision) ([]string, error) {
-		query := fmt.Sprintf(`file:(^|/)%s$ repo:^%s$@%s type:path count:all`, regexp.QuoteMeta(fileName), regexp.QuoteMeta(string(repoRev.Repo.Name)), repoRev.Commit)
+		query := fmt.Sprintf(`file:(^|/)%s$ repo:^%s$@%s type:path count:all patterntype:keyword`, regexp.QuoteMeta(fileName), regexp.QuoteMeta(string(repoRev.Repo.Name)), repoRev.Commit)
```
Specifying `patterntype:keyword` is not really necessary here, but we should always be explicit when specifying queries anyway.

stefanhengl commented on internal/batches/service/workspace_resolver.go:
```diff
@@ -387,8 +388,20 @@ func (wr *workspaceResolver) resolveRepositoriesMatchingQuery(ctx context.Contex
 
-func (wr *workspaceResolver) runSearch(ctx context.Context, query string, onMatches func(matches []streamhttp.EventMatch)) (err error) {
-	req, err := streamhttp.NewRequestWithVersion(wr.frontendInternalURL, query, searchAPIVersion)
+func determineDefaultPatternType(batchSpecVersion string) searchquery.SearchType {
+	switch batchSpecVersion {
+	case "v1":
+		return searchquery.SearchTypeStandard
+	case "v2":
+		return searchquery.SearchTypeKeyword
+	default:
+		return searchquery.SearchTypeStandard
+	}
+}
+
+func (wr *workspaceResolver) runSearch(ctx context.Context, query string, batchSpecVersion string, onMatches func(matches []streamhttp.EventMatch)) (err error) {
+	defaultPatternType := determineDefaultPatternType(batchSpecVersion)
+	req, err := streamhttp.NewRequestWithVersion(wr.frontendInternalURL, query, searchAPIVersion, &defaultPatternType)
```
This is the core of the change. The default pattern type now depends on the batch spec version.

stefanhengl commented:
> I think Julies concern is valid - this would change the behavior of existing batch specs 🤔 I'm not sure how important that property is, but it should probably at least be well called out somewhere.
 
I updated the PR and introduced a batch spec version. This ways, existing batch specs won't change behavior.

eseliger commented on lib/batches/batch_spec.go:
```diff
@@ -12,6 +13,15 @@ import (
+// <add descriptions of new versions here>
+const LatestBatchSpecVersion = "v2"
+
+var SupportedBatchSpecVersions = []string{"v1", "v2"}
```
I think we should consider integer here as the type, and `version: 2` in the spec, too - integers are easier to compare for "if version >= 2", and `version: v2` reads a bit redundantly to me as well

eseliger commented on schema/batch_spec.schema.json:
```diff
@@ -7,6 +7,10 @@
   "required": ["name"],
   "properties": {
+    "version": {
+      "type": "string",
```
could we make this an enum or add examples so the editor autocomplete shows the right versions?

eseliger commented on lib/batches/batch_spec.go:
```diff
@@ -177,6 +188,11 @@ func parseBatchSpec(schema string, data []byte) (*BatchSpec, error) {
 	}
 
+	// For backward compatibility, we allow the version field to be omitted.
+	if spec.Version != "" && !slices.Contains(SupportedBatchSpecVersions, spec.Version) {
```
you will have to update src-cli with the latest version of sg/sg/lib and probably add some logic to the query runner there, too so v2 works in CLI execution mode as well

stefanhengl replied:
@eseliger Is the src-cli change a blocker for this PR? Or can we follow-up with the src-cli change even after the release?

eseliger replied:
We usually release those in lockstep, but I don't think it has to be in for 5.5.0, we can cut a 5.5.1 as well. We should just make sure to not let that sit for too long, as batch specs shown in the web will no longer work locally (doesn't understand `version` and silently ignores it)

eseliger commented on internal/batches/service/workspace_resolver.go:
```diff
@@ -387,8 +388,20 @@ func (wr *workspaceResolver) resolveRepositoriesMatchingQuery(ctx context.Contex
+	switch batchSpecVersion {
+	case "v1":
+		return searchquery.SearchTypeStandard
+	case "v2":
+		return searchquery.SearchTypeKeyword
```
this should probably be >= v2

eseliger reviewed (approved):
Code change LGTM, left a few minor comments.

I'm a bit sad that by default, it's easy to miss that you will need to add `version: 2` to get our de-facto standard, but it seems like a lot of extra work to make heuristics based on create date here or so - so let's be pragmatic here. 

Needs to update src-cli to support this as well. 

We should also add this to the docs to make sure most people will eventually use `version: 2`.

stefanhengl commented:
> We should also add this to the docs to make sure most people will eventually use version: 2.

[Here](https://github.com/sourcegraph/docs/pull/477) is the PR for the docs.

jtibshirani commented on client/web/src/enterprise/batches/batch-spec/edit/library/comby.batch.yaml:
```diff
@@ -1,9 +1,10 @@
   in Go code with the equivalent but clearer `strconv.Iota` call.
 
-# Find all repositories that contain the `fmt.Sprintf` statement using structural search
+# Find all repositories that contain the `fmt.Sprintf` statement using regular expression search
```
Oops, thanks for fixing this after my recent PR :)

jtibshirani commented on internal/batches/service/workspace_resolver_test.go:
```diff
@@ -421,12 +421,15 @@ func resolveWorkspacesAndCompare(t *testing.T, s *store.Store, gs gitserver.Clie
-	}
-	if diff := cmp.Diff(want, have); diff != "" {
-		t.Fatalf("returned workspaces wrong. (-want +got):\n%s", diff)
+	for _, version := range []int{0, 1, 2} { // Test all versions
```
Tiny comment, it's unusual to be using "version: 0" to mean "no version is set!" Maybe we could explicitly detect the "no version" case early and swap it out for a valid version in the set `[1, 2]`?

stefanhengl replied:
I agree. The batch spec parser would be the right place it seems. However, I tried it out, and this would store the updated spec in the DB instead of the original, which causes some of the test to fail. I am not familiar enough with Batch Changes to judge if that would violate a fundamental assumption about batch specs so I reverted the change. I think that would be a nice change though, so I will follow-up once I have time to dig a bit deeper.

jtibshirani commented on CHANGELOG.md:
```diff
@@ -29,6 +29,8 @@ All notable changes to Sourcegraph are documented in this file.
 - Added syntax highlighting for the Magik programming language. [#62919](https://github.com/sourcegraph/sourcegraph/pull/62919)
 - Added syntax highlighting for the Hack programming language. [#62770](https://github.com/sourcegraph/sourcegraph/pull/62770)
+- Batch Changes: The new (optional) field "version" of batch specs determines how the spec is processed. This allows us to introduce new features while maintaining backward compatability. [#63613](https://github.com/sourcegraph/sourcegraph/pull/63613)
+- Batch Changes: A new version "v2" is introduced. Batch specs specifying `version: v2` will use keyword search as the default pattern type to determine repos/workspaces. Batch specs with `version: v1` or without version field keep using pattern type "standard". [#63613](https://github.com/sourcegraph/sourcegraph/pull/63613)
```
```suggestion
- Batch Changes: A new version `2` is introduced. Batch specs specifying `version: 2` will use keyword search as the default pattern type to determine repos/workspaces. Batch specs with `version: 1` or without version field keep using pattern type "standard". [#63613](https://github.com/sourcegraph/sourcegraph/pull/63613)
```

jtibshirani reviewed (commented):
Having a "schema version" for batch changes seems generally useful and works well for this purpose.

When someone creates a batch change through the UI, could we automatically set `version: 2`? Otherwise it's really non-obvious to users they should be doing this. And if they don't, there will be the inconsistency we're trying to avoid (defaulting to `patterntype: standard`).

stefanhengl commented:
> When someone creates a batch change through the UI, could we automatically set `version: 2`? Otherwise it's really non-obvious to users they should be doing this. And if they don't, there will be the inconsistency we're trying to avoid (defaulting to `patterntype: standard`).

That's already the case. When you create a batch change, it takes the "hello world" example as starting point and offers the other examples as alternative. I modified all of them to have the version on top.

jtibshirani commented:
> When you create a batch change, it takes the "hello world" example as starting point and offers the other examples as alternative.

Ah, right! You even explained in https://github.com/sourcegraph/sourcegraph/pull/63613#discussion_r1665745471...
